File: 14-nnfp-s1/nnfp_s1.exs
# Elixir Solutions to the first 11 problems of the 99 FP series
# Written by M.V.Harish Kumar on 14-01-2024

defmodule NNFP_S1 do
    # P01
    def last([e]), do: e
    def last([_|t]), do: last(t)

    # P02
    def penultimate([_]), do: raise "No Such element"
    def penultimate([h|[_]]), do: h
    def penultimate([_|t]), do: penultimate(t)

    # P03
    def kth([h|_], 1), do: h
    def kth([_|t], k), do: kth(t,k-1)

    # P04
    defp length([], acc), do: acc
    defp length([_|t], acc), do: length(t, acc+1)
    def length(l), do: length(l, 0)

    # P05
    defp reverse([], acc), do: acc
    defp reverse([h|t], acc), do: reverse(t, [h|acc])
    def reverse(l), do: reverse(l, [])

    # P06
    def palindrome?(l), do: l == reverse(l)

    # P07
    def flatten(list), do: flatten(list, [])
    defp flatten([h], acc), do: flatten(h, acc)
    defp flatten([h|t], acc), do: flatten(h, flatten(t, acc))
    defp flatten(n,acc), do: [n|acc]

    # P08
    def compress([]), do: []
    def compress([h|t]), do: [h | (for x <- compress(t), x != h, do: x)]

    # P09
    def pack([h|t]), do: tl pack([], [], [h|t], [h])

    def pack(final_ans, ans_ls, [curr_h | curr_t], elem) when curr_h == elem do
        pack(final_ans, ans_ls ++ [elem], curr_t, elem)
    end

    def pack(final_ans, ans_ls, [curr_h | curr_t], _) do
        pack(final_ans ++ [ans_ls], [], [curr_h | curr_t], curr_h)
    end

    def pack(final_ans, ans_ls, [], _), do: final_ans ++ [ans_ls]

    # P10
    def rle(list) do
        for packet <- pack(list), into: [], do: {Kernel.length(packet), hd packet}
    end

    # P11
    def rle_mod(list) do
        for packet <- pack(list), into: [] do
            if Kernel.length(packet) == 1 do
                hd packet
            else
                {Kernel.length(packet), hd packet}
            end
        end
    end
end

File: 14-nnfp-s1/nnfp_s1.py
# Solutions to the first 11 problems of the 99 FP series
# Written by M.V.Harish Kumar on 14-01-2024

class HList(list):
    """
    A subclass of built-in list type
    """

    def __init__(self, *iter: tuple) -> None:
        super().__init__(iter)

    def last(self) -> object:
        """ Solves P01 """
        return super().__getitem__(-1)

    def penLast(self) -> object:
        """ Solves P02 """
        return super().__getitem__(-2)

    def elementAt(self, k: int) -> object:
        """ Solves P03 """
        return super().__getitem__(k-1)

    def length(self) -> int:
        """ Solves P04 """
        l = 0
        for _ in super().__iter__(): 
            l += 1
        return l

    def reverse(self) -> list:
        """ Solves P05 """
        return super().__getitem__(slice(None, None,-1))

    def isPalindrome(self) -> bool:
        """ Solves P06 """
        return self == self.reverse()
    
    def flatten(self) -> list:
        """ Solves P07 """
        f_list = []
        for e in super().__iter__():
            if isinstance(e, list) or isinstance(e, tuple):
                f_list.extend(HList(*e).flatten())
            else:
                f_list.append(e)
        return f_list
    
    def compress(self) -> list:
        """ Solves P08 """
        prev = None
        n_list = []
        for e in super().__iter__():
            if e != prev:
                n_list.append(e)
            prev = e
        return n_list
    
    def pack(self) -> list:
        """ Solves P09 """
        n_list = [[self.elementAt(1)]]
        zipped = zip(super().__iter__(), super().__getitem__(slice(1,None)))
        for it1, it2 in zipped:
            if it1 == it2:
                n_list[-1].append(it2)
            else:
                n_list.append([it2])
        return n_list
    
    def rle(self) -> list:
        """ Solves P10 & P11 """
        s = []
        for l in self.pack():
            s.append([len(l), l[0]] if len(l) != 1 else l[0])
        return s

